### PR TITLE
LF: Make ExerciseInterface guard field optional

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
in proto.

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PhaseOne.scala [hunk 1]
```diff
@@ -718,17 +718,28 @@ private[lf] final class PhaseOne(
...
+                    )
+                  )
+                }
+              case None =>
+                Return(
+                  t.InterfaceChoiceDefRef(ifaceId, chId)(
+                    SEAbs(1, SEValue.True),
+                    cid,
+                    arg,
+                  )
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PhaseOne.scala:741]:
```suggestion
            def choiceDefRef(guard: SExpr) = 
              Return(t.InterfaceChoiceDefRef(ifaceId, chId)(
                      guard,
                      cid,
                      arg,
                    )
                  )
             maybeGuard match {
             case Some(guard) => compileExp(env, guard) choiceDefGuard(_)
             case None => choiceDefguard(SeAbs(1, SEValue.True)
             }
```
maybe something like that is a bit easier to read?

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 2]
```diff
@@ -201,7 +201,7 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
 
       val exercise_interface: Party -> ContractId M:Person -> Update Unit =
         \(exercisingParty: Party) (cId: ContractId M:Person) ->
-          Test:run @Int64 (exercise_by_interface @M:Person Nap cId 42 (\(x: M:Person) -> TRACE @Bool "interface guard" True));
+          Test:run @Int64 (exercise_interface_with_guard @M:Person Nap cId 42 (\(x: M:Person) -> TRACE @Bool "interface guard" True));
```
Comment (cocreature) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:204]: should we extend that test with something without a guard or do you rely on that being equivalent to a guard that is `const True`?

**Review by cocreature — APPROVED** [review]
Thanks!

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 3]
```diff
@@ -1386,7 +1386,11 @@ private[archive] class DecodeV1(minor: LV.Minor) {
             choice = handleInternedName(exercise.getChoiceInternedStr),
             cidE = decodeExpr(exercise.getCid, definition),
             argE = decodeExpr(exercise.getArg, definition),
-            guardE = decodeExpr(exercise.getGuard, definition),
+            guardE =
+              if (exercise.hasGuard)
+                Some(decodeExpr(exercise.getGuard, definition))
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:1391]: next PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PhaseOne.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PhaseOne.scala:741]: :+1:

**Pull request merged**